tutorials/haskell/mustache-templates/tutorial.md

```diff
@@ -0,0 +1,404 @@
+  , fiPerformer :: Text
+  , fiIndex00   :: Text
+  } deriving (Eq, Show, Generic)
+
+instance ToJSON File
+```
+
+Here we have defined two records with various fields, just like we would do
+for other domains. We also have used
+[*generics*](https://wiki.haskell.org/GHC.Generics) to quickly define
+`ToJSON` instance for the data types. You can read more about generics, but
+the technique in a nutshell is simple: compiler can derive `Generic`
+instance for your datatype (with `DeriveGeneric` extension enabled) which
+describes “structure” of your data, and then you can use that information to
+(again automatically) derive instance of `ToJSON` type class.
+
+Now we can experiment with template compilation. We will have two templates:
+`main.mustache` and `file.mustache` (we could do with just one, but then I
+wouldn't have a good example how to work with partials).
+
+The `main.mustache` looks like this:
+
+```mustache
+REM GENRE "{{& reGenre }}"
+REM DATE "{{& reDate }}"
+REM DISCID "{{& reDiscId }}"
+REM COMMENT "{{& reComment }}"
+PERFORMER "{{& rePerformer }}"
+TITLE "{{& reTitle }}"
+{{# reFiles }}
+{{> file }}
+{{/ reFiles }}
+```
+
+Here we have layout of a typical header of a CUE file. The `{{> file }}`
+part inserts a partial which we can write this way (`file.mustache` file):
+
+```mustache
+FILE "{{& fiFileName }}" WAVE
+  TRACK {{& fiIndex }} AUDIO
+    TITLE "{{& fiTitle }}"
+    PERFORMER "{{& fiPerformer }}"
+    INDEX 00 {{& fiIndex00 }}
+```
+
+Given that the template files are stored in `data/` directory, the following
+code will produce a `Template` with both of them stored in the cache (so we
+can use one from another, like we indend to do):
+
+```haskell
+main :: IO ()
+main = do
+  mainTemplate <- compileMustacheFile "data/main.mustache" -- (1)
+  fileTemplate <- compileMustacheFile "data/file.mustache" -- (2)
+  let template = mainTemplate <> fileTemplate -- (3)
+  print template -- (4)
+```
+
+What's going on?
+
+1. We load `Template` from `data/main.mustache`.
+2. We load `Template` from `data/file.mustache`.
+3. We combine caches of the two templates and get another template which is
+   just like `main` but with `file` added to its cache.
+4. We `print` the result because we don't know any better yet.
+
+Another way to do the same is to use `compileMustacheDir`:
+
+```haskell
+main :: IO ()
+main = do
+  template <- compileMustacheDir "main" "data/"
+  print template
+```
+
+Easy.
+
+### Template Haskell helpers
+
+If you do not plan to change templates “on the fly” (without re-compiling
+your application), you may like Template Haskell (TH) helpers that compile
+Mustache templates at compile time and ensure that your templates are valid.
+If some template is not valid, your program just won't compile.
+
+The TH helpers live in the `Text.Mustache.Compile.TH` module and correspond
+precisely to the three normal functions we already know. Let's play with a
+template Haskell helper right in your editing environment.
```

Comment:
“…right in your editing environment.” — I don't see what's wrong with this phrase.